lib/cartopy/mpl/geoaxes.py

```diff
@@ -1660,23 +1660,44 @@ def _pcolormesh_patched(self, *args, **kwargs):
             wrap_proj_types = (ccrs._RectangularProjection,
                                ccrs._WarpedRectangularProjection,
                                ccrs.InterruptedGoodeHomolosine,
-                               ccrs.Mercator)
+                               ccrs.Mercator,
+                               ccrs.LambertAzimuthalEqualArea,
+                               ccrs.AzimuthalEquidistant,
+                               ccrs.TransverseMercator,
+                               ccrs.Stereographic)
             if isinstance(t, wrap_proj_types) and \
                     isinstance(self.projection, wrap_proj_types):
 
                 C = C.reshape((Ny - 1, Nx - 1))
                 transformed_pts = transformed_pts.reshape((Ny, Nx, 2))
 
-                # Compute the length of edges in transformed coordinates
+                # Compute the length of diagonals in transformed coordinates
+                # This takes into account the case where the top of the cell
+                # is on one side of the boundary while the other is on the
+                # other side, case overlooked with the edges only.
+                #
+                #    Top----Top  |
+                #                |  Bottom-----Bottom
+                #
                 with np.errstate(invalid='ignore'):
-                    edge_lengths = np.hypot(
-                        np.diff(transformed_pts[..., 0], axis=1),
-                        np.diff(transformed_pts[..., 1], axis=1)
+                    ptx, pty = transformed_pts[..., 0], transformed_pts[..., 1]
+                    diagonal0_lengths = np.hypot(
+                        ptx[1:, 1:] - ptx[:-1, :-1],
+                        pty[1:, 1:] - pty[:-1, :-1]
+                    )
+                    diagonal1_lengths = np.hypot(
+                        ptx[1:, :-1] - ptx[:-1, 1:],
+                        pty[1:, :-1] - pty[:-1, 1:]
                     )
                     to_mask = (
-                        (edge_lengths > abs(self.projection.x_limits[1] -
-                                            self.projection.x_limits[0]) / 2) |
-                        np.isnan(edge_lengths)
+                        (diagonal0_lengths > (
+                            abs(self.projection.x_limits[1]
+                                - self.projection.x_limits[0])) / 2) |
```

Comment:
You have one failing test on the Goode projection, which I think is because this limit is not low enough. I'd be inclined to make this 1/3 of the map projection limits to pass that test and still be reasonably large that people probably aren't plotting 1/3 the map grid cells.

Reply:
Well the 1/2 was the original value, I didn't change it. Statistically cells with size over 1/3 would be overlapping cells and not big cells. But I didn't want to make too many changes.
We could also make this a parameter with a default value. I am not sure though how to do it without having some unforeseen implication with the signature, so I would keep the parameter for a new PR.
I would be interested in seeing the issue with the Goode projection test. I am not sure where to look though.

Reply:
I'm not sure if you can get the images off of TravisCI, but if you run the tests locally there should be a new directory `cartopy_test_output` that produces all of the images (specifically this one fails `test_mpl_integration/result-pcolormesh_goode_wrap.png`). The comparison images of what you're trying to pass for the tests are under the installation directory wherever you installed Cartopy. It looks like the upper right corner of the Goode projection is what is different.

The 1/3 limit will only add more cells to the `pcolor` calculations, but those cells are likely more correct due to the polygon clipping I mentioned earlier, so I don't think there is a major downside to changing this to 1/3.